packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,489 @@
+    if (this.state.allNonEvmTokens[accountId]) {
+      this.update((state) => {
+        delete state.allNonEvmTokens[accountId];
+      });
+    }
+  }
+
+  /**
+   * Refreshes the assets snaps and metadata for the given list of assets
+   *
+   * @param assets - The assets to refresh
+   */
+  async #refreshAssetsMetadata(assets: CaipAssetType[]) {
+    const assetsWithoutMetadata: CaipAssetType[] = assets.filter(
+      (asset) => !this.state.metadata[asset],
+    );
+
+    // call the snap to get the metadata
+    if (assetsWithoutMetadata.length > 0) {
+      // check if for every asset in assetsWithoutMetadata there is a snap in snaps by chainId else call getAssetSnaps
+      if (
+        !assetsWithoutMetadata.every((asset: CaipAssetType) => {
+          const chainIdResponse = parseCaipAssetType(asset);
+          return Boolean(this.#getAssetSnapFor(chainIdResponse.chainId));
+        })
+      ) {
+        this.#snaps = this.#getAssetSnaps();
+      }
+      await this.#updateAssetsMetadata(assetsWithoutMetadata);
+    }
+  }
+
+  /**
+   * Updates the assets metadata for the given list of assets
+   *
+   * @param assets - The assets to update
+   */
+  async #updateAssetsMetadata(assets: CaipAssetType[]) {
+    // Creates a mapping of scope to their respective assets list.
+    const assetsByScope: Record<CaipChainId, CaipAssetType[]> = {};
+    for (const asset of assets) {
+      const chainIdResponse = parseCaipAssetType(asset);
+      if (!assetsByScope[chainIdResponse.chainId]) {
+        assetsByScope[chainIdResponse.chainId] = [];
+      }
+      assetsByScope[chainIdResponse.chainId].push(asset);
+    }
+    let newMetadata: Record<CaipAssetType, FungibleAssetMetadata> = {};
+    for (const chainId in assetsByScope) {
+      if (hasProperty(assetsByScope, chainId)) {
+        const assetsForChain = assetsByScope[chainId as CaipChainId];
+        // Now fetch metadata from the associated asset Snaps:
+        const snap = this.#getAssetSnapFor(chainId as CaipChainId);
+        if (snap) {
+          const metadata = await this.#getMetadata(assetsForChain, snap.id);
+          newMetadata = {
+            ...newMetadata,
+            ...(metadata ? metadata.assets : {}),
+          };
+        }
+      }
+    }
+    this.update((state) => {
+      state.metadata = {
+        ...this.state.metadata,
+        ...newMetadata,
+      };
+    });
+  }
+
+  /**
+   * Creates a mapping of CAIP-2 Chain ID to Asset Snaps.
+   *
+   * @returns A mapping of CAIP-2 Chain ID to Asset Snaps.
+   */
+  #getAssetSnaps(): Record<CaipChainId, Snap[]> {
+    const snaps: Record<CaipChainId, Snap[]> = {};
+    const allSnaps = this.#getAllSnaps();
+    const allPermissions = allSnaps.map((snap) =>
+      this.#getSnapsPermissions(snap.id),
+    );
+
+    for (const [index, permission] of allPermissions.entries() as unknown as [
+      number,
+      AssetEndowment,
+    ][]) {
```

Comment:
Is the type-cast really necessary here?

Reply:
Fixed here [d4cd884](https://github.com/MetaMask/core/pull/5138/commits/d4cd88440f16d22c1152f46cc7a8d0a52df11403)